tket/src/Circuit/DAGProperties.cpp

```diff
@@ -107,31 +122,121 @@ bool is_valid(const DAG &G) {
-      CHECK(c_out_ports.size() == out_deg);
-      CHECK(
-          (in_deg == 0 && out_deg == 1) || (in_deg == 1 && out_deg == 0) ||
-          (c_in_ports == c_out_ports));  // bijection between in and out ports
-    } else {
-      // Check that it is a Measure vertex.
+    // get size and empty only once
+
+    unsigned q_in_deg = q_in.size();
+    unsigned q_out_deg = q_out.size();
+
+    unsigned c_in_deg = c_in.size();
+    unsigned c_out_deg = c_out.size();
+
+    unsigned b_in_deg = b_in.size();
+
+    unsigned w_in_deg = w_in.size();
+    unsigned w_out_deg = w_out.size();
+
+    bool q_in_empty = q_in.empty();
+    bool q_out_empty = q_out.empty();
+
+    bool c_in_empty = c_in.empty();
+    bool c_out_empty = c_out.empty();
+
+    bool b_in_empty = b_in.empty();
+    bool b_out_empty = b_out.empty();
+
+    bool w_in_empty = w_in.empty();
+    bool w_out_empty = w_out.empty();
+
+    // check that ports maps to edges
+
+    CHECK(q_in_ports.size() == q_in_deg);
+    CHECK(q_out_ports.size() == q_out_deg);
+
+    CHECK(w_in_ports.size() == w_in_deg);
+    CHECK(w_out_ports.size() == w_out_deg);
+
+    CHECK(c_in_ports.size() == c_in_deg);
+    CHECK(c_out_ports.size() == c_out_deg);
+
+    CHECK(b_in_ports.size() == b_in_deg);
+
+    if (c_in_empty && c_out_empty && q_in_empty && !q_out_empty) {
+      // quantum in vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_deg == 0 && q_out_deg == 1);
+    } else if (c_in_empty && c_out_empty && !q_in_empty && q_out_empty) {
+      // quantum out vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_deg == 1 && q_out_deg == 0);
+    } else if (c_in_empty && c_out_empty && !q_in_empty && !q_out_empty) {
+      // quantum vertex
+      CHECK(b_out_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_ports == q_out_ports);  // bijection between in and out ports
+    } else if (w_in_empty && !w_out_empty) {
+      // wasm out vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(c_in_empty);
+      CHECK(c_out_empty);
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(w_in_deg == 0 && w_out_deg == 1);
+    } else if (!w_in_empty && w_out_empty) {
+      // wasm in vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(c_in_empty);
+      CHECK(c_out_empty);
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(w_in_deg == 1 && w_out_deg == 0);
+    } else if (!w_in_empty && !w_out_empty) {
+      // wasm vertex
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(c_in_ports == c_out_ports);  // bijection between in and out ports
+      CHECK(w_in_ports == w_out_ports);  // bijection between in and out ports
+    } else if (c_in_empty && !c_out_empty) {
+      // claccical in vertex, can have bool out edges
```

Comment:
```suggestion
      // classical in vertex, can have bool out edges
```

Reply:
done in [suggested changes I](https://github.com/CQCL/tket/pull/743/commits/e06543a3343027c8ff962d77fda395d3b22d9f96)  [e06543a](https://github.com/CQCL/tket/pull/743/commits/e06543a3343027c8ff962d77fda395d3b22d9f96)